### PR TITLE
[REF] mail, *: use wowl env instead of legacy env

### PR DESCRIPTION
**Pull request by alexkuhn**
*: hr, hr_holidays, im_livechat, sms, snailmail, web,
   website_livechat, website_slides

- passing WOWL env to DiscussComponent and ChatterContainer is crappy
- todo: functional test with all modules
- todo: adapt all qunit tests

Task-2582313

enterprise PR: https://github.com/odoo/enterprise/pull/19286

### CODE REVIEW

File: addons/mail/static/src/services/messaging_service/messaging_service.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        env.bus.on(
+            'will_hide_home_menu',
+            null,
+            () => this.messagingBus.trigger('will_hide_home_menu')
+        );
+        env.bus.on(
+            'will_show_home_menu',
+            null,
+            () => this.messagingBus.trigger('will_show_home_menu')
+        );
```
Comment (seb-odoo) [inline addons/mail/static/src/services/messaging_service/messaging_service.js:93]: There must be a better way to do this with the new home menu service / events.

File: addons/mail/static/src/services/messaging_service/messaging_service.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        });
+        /**
+         * Registry of models.
+         */
+        this.models = {};
+        /**
+         * Environment keys used in messaging.
+         */
+        Object.assign(this, {
+            autofetchPartnerImStatus: true,
```
Comment (seb-odoo) [inline addons/mail/static/src/services/messaging_service/messaging_service.js:31]: move as field on messaging model?

File: addons/mail/static/src/services/messaging_service/messaging_service.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            destroyMessaging() {
+                if (this.modelManager) {
+                    this.modelManager.deleteAll();
+                    this.messaging = undefined;
+                }
+            },
+            disableAnimation: false,
```
Comment (seb-odoo) [inline addons/mail/static/src/services/messaging_service/messaging_service.js:38]: move as field on messaging model?

File: addons/mail/static/src/services/messaging_service/messaging_service.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                return this.messaging.isInitialized;
+            },
+            /**
+             * States whether the environment is in QUnit test or not.
+             *
+             * Useful to prevent some behaviour in QUnit tests, like applying
+             * style of attachment that uses url.
+             */
+            isQUnitTest: false,
+            loadingBaseDelayDuration: 400,
```
Comment (seb-odoo) [inline addons/mail/static/src/services/messaging_service/messaging_service.js:52]: move as fields on messaging model?

File: addons/mail/static/src/services/messaging_service/messaging_service.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            messagingBus: new EventBus(),
+            /**
+             * Promise which becomes resolved when messaging is created.
+             *
+             * Useful for discuss widget to know when messaging is created, because this
+             * is an essential condition to make it work.
+             */
+            messagingCreatedPromise: this.createMessaging(),
+            modelManager: new ModelManager(env),
+            store,
```
Comment (seb-odoo) [inline addons/mail/static/src/services/messaging_service/messaging_service.js:63]: should be a messaging_store service instead?

File: addons/mail/static/src/services/messaging_service/messaging_service.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        const store = new Store({
+            env,
+            state: {
+                messagingRevNumber: 0,
+            },
+        });
+        /**
+         * Registry of models.
+         */
+        this.models = {};
```
Comment (seb-odoo) [inline addons/mail/static/src/services/messaging_service/messaging_service.js:26]: possible to create a messaging_model service instead?

File: addons/mail/static/src/services/messaging_service/messaging_service.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            loadingBaseDelayDuration: 400,
+            messaging: undefined,
+            messagingBus: new EventBus(),
+            /**
+             * Promise which becomes resolved when messaging is created.
+             *
+             * Useful for discuss widget to know when messaging is created, because this
+             * is an essential condition to make it work.
+             */
+            messagingCreatedPromise: this.createMessaging(),
```
Comment (seb-odoo) [inline addons/mail/static/src/services/messaging_service/messaging_service.js:61]: Is it still necessary? Possible to use the service booting code to wait on this without a custom promise? Maybe with a new messaging_boot service that would be different than messaging itself?

**Review by seb-odoo — COMMENTED** [review]
First check, overall the main issue I have currently is the indirection/repetition when accessing messaging keys on the env.

I think we should find a way to have service.messaging be the same messaging as before (the record), and move the other keys that were in the env before in a service of their own, instead of as part of messaging service.
Below some suggestions in that direction, but I'm not sure all of them are directly applicable.

**Comment by seb-odoo** [comment]
Also need to rebase to fix conflicts on the export default PR.

File: addons/mail/views/mail_channel_views.xml [hunk 8]
```diff
@@ -145,9 +145,9 @@
 
     <record id="action_discuss" model="ir.actions.client">
         <field name="name">Discuss</field>
-        <field name="tag">mail.widgets.discuss</field>
+        <field name="tag">mail.discuss</field>
         <field name="res_model">mail.channel</field>
-        <field name="params" eval="&quot;{ 'default_active_id': 'mail.box_inbox' }&quot;"/>
+        <field name="default_active_id">mail.box_inbox</field>
```
Comment (seb-odoo) [inline addons/mail/views/mail_channel_views.xml:150]: need to revert this change to fix issue?

File: addons/mail/static/src/components/discuss/discuss.js [hunk 9]
```diff
@@ -54,29 +56,39 @@ export class Discuss extends Component {
...
+        if (this.props.action.activeId) {
+            this.discuss.update({
+                initActiveId: this.props.action.activeId,
+            });
+        }
         this.discuss.update({ isOpen: true });
         if (this.discuss.thread) {
-            this.trigger('o-push-state-action-manager');
-        } else if (this.env.isMessagingInitialized()) {
+            // TODO: update action active id with discuss active id
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/discuss.js:71]: pushState in router service

File: addons/hr/static/src/models/employee/employee.js [hunk 10]
```diff
@@ -55,18 +55,20 @@ function factory(dependencies) {
          * @param {integer[]} param0.ids
          */
         static async performRpcRead({ context, fields, ids }) {
-            const employeesData = await this.env.services.rpc({
-                model: 'hr.employee.public',
-                method: 'read',
-                args: [ids],
-                kwargs: {
+            const employeesData = await this.env.services.rpc(
```
Comment (seb-odoo) [inline addons/hr/static/src/models/employee/employee.js:58]: orm?

File: addons/mail/static/src/widgets/messaging_menu/messaging_menu.js [hunk 11]
```diff
@@ -1,50 +1,26 @@
...
 
-export default MessagingMenuWidget;
+const systrayItem = {
+    Component: MessagingMenu,
+};
+registry.category('systray').add(
+    'mail.messaging_menu',
+    systrayItem,
+    { sequence: 10 },
+);
```
Comment (seb-odoo) [inline addons/mail/static/src/widgets/messaging_menu/messaging_menu.js:26]: The register of messaging menu component should not be on the old widget file, but on the boot file/messaging service file.

**Review by seb-odoo — CHANGES_REQUESTED** [review]
I tried my best to review ignoring the current namespacing issue/insane noise that this introduces.

See the comments below.

The next step would be to make runbot green. At least with problematic tests skipped or commented, so that we can see what's left to do.

File: addons/mail/static/src/components/thread_view/thread_view.js [hunk 12]
```diff
@@ -99,22 +97,13 @@ export class ThreadView extends Component {
...
     // Private
     //--------------------------------------------------------------------------
 
-    /**
-     * Called when thread component is mounted or patched.
-     *
-     * @private
-     */
-    _update() {
-        this.trigger('o-rendered');
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/components/thread_view/thread_view.js:115]: Was used to update control panel to match breadcrumb with active thread. Upcoming discuss topbar should rely on model so there is no need for replacement.

File: addons/mail/static/src/js/core/translation.js [hunk 13]
```diff
@@ -1,25 +0,0 @@
...
-        if (Component.env.messaging) {
-            // Update messaging locale whenever the translation bundle changes.
-            // In particular if messaging is created before the end of the
-            // `load_translations` RPC, the default values have to be
-            // updated by the received ones.
-            Component.env.messaging.locale.update({
-                language: this.parameters.code,
-                textDirection: this.parameters.direction,
-            });
-        }
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/js/core/translation.js:22]: legacy translation should now rely on services, especially `user` and `localization` services

File: addons/mail_bot/static/src/models/messaging_initializer/messaging_initializer.js [hunk 14]
```diff
@@ -11,14 +11,14 @@ registerInstancePatchModel('mail.messaging_initializer', 'mail_bot/static/src/mo
...
-        }));
+        const data = await this.async(() => this.env.services.orm.call(
+            'mail.channel',
+            'init_odoobot',
+        ));
         if (!data) {
             return;
         }
-        this.env.session.odoobot_initialized = true;
+        odoo.session_info.odoobot_initialized = true;
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail_bot/static/src/models/messaging_initializer/messaging_initializer.js:21]: This worked before but not anymore. The data comes from overridden python method `session_info`, but I don't know how to retrieve from new `user` service in some ways. Or another way is move that into an override of `init_messaging` of some sort... What do you think?

File: addons/mail/static/src/components/discuss/tests/discuss_moderation_tests.js [hunk 15]
```diff
@@ -1 +0,0 @@
-
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/components/discuss/tests/discuss_moderation_tests.js:1]: yes

**Comment by rdeodoo** [comment]
According to the (deployed) task, this seems to have been done with https://github.com/odoo/odoo/pull/83774
